code/modules/hydroponics/compost_bin.dm

```diff
@@ -0,0 +1,195 @@
+#define MIN_CONVERSION 10
+/**
+  * # compost bin
+  * used to make soil from plants.
+  * Doesn't have components.
+  */
+/obj/machinery/compost_bin
+	name = "compost bin"
+	desc = "A wooden bin for composting."
+	icon = 'icons/obj/hydroponics/equipment.dmi'
+	icon_state = "compost_bin-empty"
+	power_state = NO_POWER_USE
+	density = TRUE
+	anchored = TRUE
+	/// amount of biomass in the compost bin
+	var/biomass = 0
+	/// amount of compost in the compost bin
+	var/compost = 0
+	/// The maximum amount of biomass the compost bin can store.
+	var/biomass_capacity = BASE_BIOMASS_CAPACITY
+	/// The maximum amount of compost the compost bin can store.
+	var/compost_capacity = BASE_COMPOST_CAPACITY
+	var/composting = FALSE
+
+/obj/machinery/compost_bin/Initialize(mapload)
+	/// try to compost
+	compost()
+	return ..()
+
+
+/obj/machinery/compost_bin/on_deconstruction()
+	/// returns wood instead of the non-existent components
+	new /obj/item/stack/sheet/wood(loc, 10)
+	..()
+
+/obj/machinery/compost_bin/screwdriver_act(mob/living/user, obj/item/I)
+	/// there are no screws either
+	to_chat(user, "<span class='warning'>[src] has no screws!</span>")
+
+/obj/machinery/compost_bin/crowbar_act(mob/living/user, obj/item/I)
+	/// no panel either
+	return default_deconstruction_crowbar(user, I, ignore_panel = TRUE)
+
+/// takes care of plant insertion and conversion to biomass, and start composting what was inserted
+/obj/machinery/compost_bin/attackby(obj/item/O, mob/user, params)
+	if(user.a_intent == INTENT_HARM)
+		return ..()
+
+	else if(istype(O, /obj/item/storage/bag/plants))
+		if(biomass >= biomass_capacity)
+			to_chat(user, "<span class='warning'>[src] can't hold any more biomass!</span>")
+			return
+
+		var/obj/item/storage/bag/plants/PB = O
+		for(var/obj/item/reagent_containers/food/snacks/grown/G in PB.contents)
+
+			if(biomass >= biomass_capacity)
+				break
+
+			/// create biomass from plant nutriment and plant matter
+			var/plant_biomass = G.reagents.get_reagent_amount("nutriment") + G.reagents.get_reagent_amount("plantmatter")
+
+			biomass += clamp(plant_biomass*10, 1 ,biomass_capacity-biomass)
+
+			PB.remove_from_storage(G, src)
+
+		/// start composting after plants are inserted
+		compost()
+
+		if(biomass < biomass_capacity)
+			to_chat(user, "<span class='info'>You empty [PB] into [src].</span>")
+		else
+			to_chat(user, "<span class='info'>You fill [src] to its capacity.</span>")
+
+		SStgui.update_uis(src)
+		return TRUE
+
+	else if(istype(O, /obj/item/reagent_containers/food/snacks/grown))
+		if(biomass >= biomass_capacity)
+			to_chat(user, "<span class='warning'>[src] can't hold any more plants!</span>")
+			return
+		if(!user.unEquip(O))
+			return
+
+		O.forceMove(src)
+		/// make biomass from nutriment and plant matter
+		var/plant_biomass = O.reagents.get_reagent_amount("nutriment") + O.reagents.get_reagent_amount("plantmatter")
+		biomass += clamp(plant_biomass*10, 1 ,biomass_capacity-biomass)
+		qdel(O)
```

Comment:
Can you make this into a separate proc for putting in any grown food? Something like
```
/obj/machinery/compost_bin/proc/add_grown(obj/item/reagent_containers/food/snacks/grown)
	O.forceMove(src)
	/// make biomass from nutriment and plant matter
	var/plant_biomass = O.reagents.get_reagent_amount("nutriment") + O.reagents.get_reagent_amount("plantmatter")
		biomass += clamp(plant_biomass*10, 1 ,biomass_capacity-biomass)
	qdel(O)

Reply:
yeah sure

Reply:
Though I think I'll only make a separate function for the insertion and conversion to biomass, because stuff from the plant bag is treated a bit differently in how it's removed.